ipq40xx: add support for EnGenius EMD1

Yen-Ting-Shen opened the pull request:
SOC:     IPQ4018 / QCA Dakota
        CPU:     Quad-Core ARMv7 Processor rev 5 (v7l) Cortex-A7
        DRAM:    256 MiB
        NOR:     32 MiB
        ETH:     Qualcomm Atheros QCA8072 (1 port)
        WLAN1:   Qualcomm Atheros QCA4018 2.4GHz 802.11bgn 2:2x2
        WLAN2:   Qualcomm Atheros QCA4018 5GHz 802.11a/n/ac 2:2x2
        INPUT:   RESET Button
        LEDS:    White, Blue, Red, Orange

        Flash instruction:

        From EnGenius firmware to OpenWrt firmware:

        In Firmware Upgrade page, upgrade your openwrt-ipq40xx-generic-engenius_emd1-squashfs-factory.bin directly.

        From OpenWrt firmware to EnGenius firmware:

        1. Setup a TFTP server on your computer and configure static IP to 192.168.99.8
           Put the EnGenius firmware in the TFTP server directory on your computer.
        2. Power up EMD1. Press 4 and then press any key to enter u-boot.
        3. Download EnGenius firmware
           (IPQ40xx) # tftpboot 0x84000000 openwrt-ipq40xx-emd1-nor-fw-s.img
        4. Flash the firmware
           (IPQ40xx) # imgaddr=0x84000000 && source 0x84000000:script
        5. Reboot
           (IPQ40xx) # reset

Signed-off-by: Yen-Ting-Shen <frank.shen@senao.com>

Thanks for your contribution to OpenWrt!

To help keep the codebase consistent and readable,
and to help people review your contribution,
we ask you to follow the rules you find in the wiki at this link
https://openwrt.org/submitting-patches

Please remove this message before posting the pull request.

adrianschmutzler commented:
Does this device have a MAC address on label/sticker/case/box?

Yen-Ting-Shen commented:
> Does this device have a MAC address on label/sticker/case/box?

Yes, a MAC address for LAN is on the label.

adrianschmutzler commented:
Which is it?
$(mtd_get_mac_ascii 0:APPSBLENV ethaddr) or $(mtd_get_mac_ascii 0:APPSBLENV wlanaddr)

You can add the the corresponding address to 02_network then by adding
label_mac=lan_mac
or
label_mac=$(mtd_get_mac_ascii 0:APPSBLENV wlanaddr)
as I did in this patch: https://patchwork.ozlabs.org/patch/1166089/

Yen-Ting-Shen commented:
It's $(mtd_get_mac_ascii 0:APPSBLENV ethaddr).
updated. Thanks for your reminder.

adrianschmutzler commented:
Hi, you need to rebase your changes onto recent master.

Despite, you have indent in your commit message that should be removed.

Yen-Ting-Shen commented:
Updated the commit message and rebased this PR, thanks for your reminder.

chunkeey commented on target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-emd1.dts:
```diff
@@ -0,0 +1,219 @@
+		};
+
+		red {
+			label = "emd1:red";
```
See:

https://github.com/torvalds/linux/blob/master/Documentation/leds/leds-class.rst#led-device-naming

> LED Device Naming
>
> Is currently of the form:
>  "devicename:color:function"

From what I can tell, your label lacks the "function".

Yen-Ting-Shen replied:
I'll rebase the code later.

chunkeey commented on target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-emd1.dts:
```diff
@@ -0,0 +1,219 @@
+	status = "okay";
+	cs-gpios = <&tlmm 54 GPIO_ACTIVE_HIGH>;
+
+	m25p80@0 {
```
wouldn't "flash@0" work as well?

Yen-Ting-Shen replied:
It works.
Should I use "flash@0" instead of "m25p80@0"?

adrianschmutzler replied:
If it works, yes.

chunkeey commented on target/linux/ipq40xx/base-files/etc/board.d/02_network:
```diff
@@ -59,6 +59,9 @@ ipq40xx_setup_interfaces()
 	compex,wpj428)
 		ucidef_set_interface_lan "eth0 eth1"
 		;;
+	engenius,emd1)
```
Hm, this looks odd. From what I can deduce, you have a "unused" eth0 in your system, right?

Can't you just edit the dts and do

    soc {
          ....

                edma@c080000 {
                        status = "okay";
                        qcom,num_gmac = <1>;
                };
    };

    &gmac0 {
        vlan_tag = <1 0x20>;
    };

and add the "engenius,emd1" to the ```ucidef_set_interface_lan "eth0"``` list to get the same result with "eth0" now being the only ethernet adapter?

Yen-Ting-Shen replied:
Yes, you are right!
Thanks for your suggestion.
I've updated the codes.
Please have a review again.

Yen-Ting-Shen replied:
I have only one "eth0" now.

chunkeey replied:
I copied that from #2398 . The question is, if we should make this properly and also change the switchbitmaps and add the poll-link properties. Because from what I remember, your device  currently doesn't notify netifd when the cable gets unplugged and replugged/when the link gets downgraded.

I think the following would "work", but it needs testing (see the inline comment)

         soc {
                         ess-switch@c000000 {
                                 switch_lan_bmp = <0x20>;
                                 switch_wan_bmp = <0x00>; /* not sure without HW to test, if 0x00 works, if not I would accept 0x1e here */
                          };
         };

         &gmac0 {
                 local-mac-address = [00 00 00 00 00 00];
                 qcom,phy_mdio_addr = <4>;
                 qcom,poll_required = <1>;
                 qcom,forced_speed = <1000>;
                 qcom,forced_duplex = <1>;
                 vlan_tag = <1 0x20>;
         };

(In theory, you could also add a ```ucidef_add_switch "switch0" "0u@eth0" "5:lan"```, but this might end up just confusing people.)

Yen-Ting-Shen replied:
I've updated the "dts" file as you mentioned.
The ethernet works fine, and I can see kernel shows "eth0: GMAC Link is up with phy_speed=1000" and "eth0: GMAC Link is down" when I plug and unplug the cable.
But how do I check that it notifies netifd normally?

dedeckeh replied:
You can check the carrier status of a device in netifd by using the ubus call network.device status

Yen-Ting-Shen replied:
Thanks to dedeckeh! The command is useful.
The device notifies netifd link status normally, please have a review again.

adrianschmutzler commented on target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-emd1.dts:
```diff
@@ -0,0 +1,226 @@
+			gpios = <&tlmm 58 GPIO_ACTIVE_LOW>;
+		};
+
+		wlan2G {
```
I'd prefer to have those lower-case only (and the label as well): wlan2g and wlan5g

Yen-Ting-Shen replied:
done.

adrianschmutzler commented on target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-emd1.dts:
```diff
@@ -0,0 +1,226 @@
+	leds {
+		compatible = "gpio-leds";
+
+		power: white {
```
The node name should be "power" as well and I prefer to have the node label with "led_" prefix, so you get:
`led_power: power {`
Obviously, you will need to adjust aliases to "led_power" as well.

Yen-Ting-Shen replied:
updated.
Thanks for your reminder.

adrianschmutzler commented on target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-emd1.dts:
```diff
@@ -0,0 +1,226 @@
+			drive-strength = <12>;
+			bias-disable;
+		};
+		pin_cs {
```
Please add an empty line between blocks.

Yen-Ting-Shen replied:
done.

adrianschmutzler commented on target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-emd1.dts:
```diff
@@ -0,0 +1,226 @@
+			partition9@200000 {
+				compatible = "denx,fit";
+				label = "firmware";
+				reg = <0x200000 0x1e00000>;
```
Though this is not really important: This is the only entry where you do not use eight-digit addresses/sizes.
I'd either change this one to 8-digit, too, or remove the useless leading zeros for the other partitions.

Yen-Ting-Shen replied:
I've changed this one to 8-digit.
Thanks for your suggestion.

adrianschmutzler commented on target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-emd1.dts:
```diff
@@ -0,0 +1,226 @@
+				reg = <0x00000000 0x00040000>;
+				read-only;
+			};
+			partition1@40000 {
```
I'd prefer to have the numbers before the "@" removed as they have no use and the common scheme is "partition@start"

chunkeey replied:
hmpf. Since I added those to one of my devices, let me explain why I do like them.

These partition$X "numbers" are great if you start from a /proc/mtd + bootlog dump and have to write the .dts node entries from scratch. The idea there is that the "/dev/mtd$X" matches the $X in the partition$X@offset, this way you can verify bootlogs/dumps/faillogs and see if you've missed something or made an error.

adrianschmutzler replied:
Okay, if there is an actual use to that, I won't play the showstopper here.

Note that we advise otherwise in https://openwrt.org/docs/guide-developer/defining-firmware-partitions . However, that rule could be weakened there.

I'd then change the first "partition@" to "partition0@" for this device. We just have to take care that we won't see too much copy/paste errors for other devices then (I've once seen three partition5@ in a row ...).

Yen-Ting-Shen replied:
I've changed the "partition@" to "partition0@".
Thanks for the explanation and suggestion.

adrianschmutzler commented on target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-emd1.dts:
```diff
@@ -0,0 +1,234 @@
+		led-failsafe = &led_power;
+		led-running = &led_power;
+		led-upgrade = &led_power;
+	};
```
Don't know whether that also applies to this target, but I would expect the aliases block at the beginning directly after model/compatible.

Yen-Ting-Shen replied:
I've moved it up.

blocktrron commented on target/linux/ipq40xx/image/Makefile:
```diff
@@ -241,6 +241,21 @@ define Device/engenius_eap1300
 TARGET_DEVICES += engenius_eap1300
 
+define Device/engenius_emd1
+	$(call Device/FitImage)
```
Does the bootloader support LZMA packed kernels? If so, it's preferred to use an LZMA packed kernel image.

Yen-Ting-Shen replied:
The bootloader doesn't support LZMA compression.

blocktrron commented on target/linux/ipq40xx/image/Makefile:
```diff
@@ -241,6 +241,21 @@ define Device/engenius_eap1300
+	DEVICE_DTS_CONFIG := config@4
+	BOARD_NAME := emd1
+	SOC := qcom-ipq4018
+	KERNEL_SIZE := 4096k
```
Is this a hard limit? Fro my experience, >4MB large kernels can be loaded from SPI flash without a problem.

As it's not used anywhere, you should consider dropping the kernel size definition.

Yen-Ting-Shen replied:
It's not used here. I've removed it.

Yen-Ting-Shen committed:
ipq40xx: add support for EnGenius EMD1

SOC:     IPQ4018 / QCA Dakota
CPU:     Quad-Core ARMv7 Processor rev 5 (v7l) Cortex-A7
DRAM:    256 MiB
NOR:     32 MiB
ETH:     Qualcomm Atheros QCA8072 (1 port)
WLAN1:   Qualcomm Atheros QCA4018 2.4GHz 802.11bgn 2:2x2
WLAN2:   Qualcomm Atheros QCA4018 5GHz 802.11a/n/ac 2:2x2
INPUT:   RESET Button
LEDS:    White, Blue, Red, Orange

Flash instruction:

From EnGenius firmware to OpenWrt firmware:

In Firmware Upgrade page, upgrade your openwrt-ipq40xx-generic-engenius_emd1-squashfs-factory.bin directly.

From OpenWrt firmware to EnGenius firmware:

1. Setup a TFTP server on your computer and configure static IP to 192.168.99.8
   Put the EnGenius firmware in the TFTP server directory on your computer.
2. Power up EMD1. Press 4 and then press any key to enter u-boot.
3. Download EnGenius firmware
   (IPQ40xx) # tftpboot 0x84000000 openwrt-ipq40xx-emd1-nor-fw-s.img
4. Flash the firmware
   (IPQ40xx) # imgaddr=0x84000000 && source 0x84000000:script
5. Reboot
   (IPQ40xx) # reset

Signed-off-by: Yen-Ting-Shen <frank.shen@senao.com>

chunkeey commented:
Merged! Thanks.
